### PR TITLE
VP-2696 CRUD for rules preview

### PR DESCRIPTION
**Pull request by AliveMen**
Add searchService and service
Add controller
Implement search, save, get

### CODE REVIEW

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        /// <summary>
+        /// Create/Update the specified association.
+        /// </summary>
+        /// <param name="DynamicAssociation">The dynamic association rule.</param>
+        [HttpPost]
+        [Route("")]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status200OK)]
+        public async Task<ActionResult<CatalogProduct>> SaveProduct([FromBody] DynamicAssociation association)
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:52]: SaveAssociations(DynamicAssociation[] association)

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            //var authorizationResult = await _authorizationService.AuthorizeAsync(User, association, new CatalogAuthorizationRequirement(ModuleConstants.Security.Permissions.Update));
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:54]: Probably this check should be uncommented

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            //if (!authorizationResult.Succeeded)
+            //{
+            //    return Unauthorized();
+            //}
+
+            await _service.SaveChangesAsync(new[] { association });
+                
+            return NoContent();
+        }
+    }
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:64]: Please add delete endpoint ([FromQuery]string[] ids)

File: src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.CatalogModule.Core.Model.Search
+{
+    public class DynamicAssociationSearchCriteria : SearchCriteriaBase
+    {
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs:6]: Please add StoreIds and Groups to the search criteria, and use them in the search service query

File: src/VirtoCommerce.CatalogModule.Data/Search/DynamicAssociationSearchService.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+            return query;
+        }
+
+        protected virtual IList<SortInfo> BuildSortExpression(DynamicAssociationSearchCriteria criteria)
+        {
+            var sortInfos = criteria.SortInfos;
+            if (sortInfos.IsNullOrEmpty())
+            {
+                sortInfos = new[] {new SortInfo {SortColumn = nameof(ItemEntity.Name)}};
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Data/Search/DynamicAssociationSearchService.cs:89]: DynamicAssociation.Name,

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs (hunk 1 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs]: renamed

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs]: done

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs (hunk 3 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:91]: added

File: src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs (hunk 4 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs:6]: done

File: src/VirtoCommerce.CatalogModule.Data/Search/DynamicAssociationSearchService.cs (hunk 5 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Data/Search/DynamicAssociationSearchService.cs]: fixed

File: src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs (hunk 4 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs:6]: Please add them as string arrays

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        /// <summary>
+        /// Create/Update the specified association.
+        /// </summary>
+        /// <param name="association">The dynamic association rule.</param>
+        [HttpPost]
+        [Route("")]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status200OK)]
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:55]: Seems it is redundant

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        public async Task<ActionResult<CatalogProduct>> SaveAssociations([FromBody] DynamicAssociation association)
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:56]: DynamicAssociation[] associations

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        {
+            var authorizationResult = await _authorizationService.AuthorizeAsync(User, association, new CatalogAuthorizationRequirement(ModuleConstants.Security.Permissions.Update));
+            if (!authorizationResult.Succeeded)
+            {
+                return Unauthorized();
+            }
+
+            await _service.SaveChangesAsync(new[] { association });
+                
+            return NoContent();
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:66]: Need to return saved associations

File: src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs (hunk 4 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Core/Model/Search/DynamicAssociationSearchCriteria.cs:6]: Same for the groups

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs [hunk 9]
```diff
@@ -46,24 +47,25 @@ public async Task<ActionResult<DynamicAssociation>> GetAssociationById(string id
         }
 
         /// <summary>
-        /// Create/Update the specified association.
+        /// Create/Update the array of association.
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:50]: /// Create/Update associations

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs (hunk 6 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleDynamicAssociationsController.cs:55]: [ProducesResponseType(typeof(void), StatusCodes.Status200OK)]
 could be removed too (based on ActionResult tyoe)

**Pull request merged**